runtime/service/src/test/java/org/apache/polaris/service/events/PolarisEventTest.java

```diff
@@ -0,0 +1,109 @@
+
+  @Test
+  void testBuilderCreatesEvent() {
+    PolarisEvent event =
+        PolarisEvent.builder(PolarisEventType.BEFORE_CREATE_TABLE, TEST_METADATA)
+            .attribute(EventAttributes.CATALOG_NAME, "my-catalog")
+            .attribute(EventAttributes.TABLE_NAME, "my-table")
+            .build();
+
+    assertThat(event.type()).isEqualTo(PolarisEventType.BEFORE_CREATE_TABLE);
+    assertThat(event.metadata()).isEqualTo(TEST_METADATA);
+    assertThat(event.attribute(EventAttributes.CATALOG_NAME)).isEqualTo(Optional.of("my-catalog"));
+    assertThat(event.attribute(EventAttributes.TABLE_NAME)).isEqualTo(Optional.of("my-table"));
+  }
+
+  @Test
+  void testAttributeReturnsEmptyForMissingKey() {
+    PolarisEvent event =
+        PolarisEvent.builder(PolarisEventType.BEFORE_CREATE_TABLE, TEST_METADATA).build();
+
+    assertThat(event.attribute(EventAttributes.CATALOG_NAME)).isEmpty();
+  }
+
+  @Test
+  void testHasAttribute() {
+    PolarisEvent event =
+        PolarisEvent.builder(PolarisEventType.BEFORE_CREATE_TABLE, TEST_METADATA)
+            .attribute(EventAttributes.CATALOG_NAME, "my-catalog")
+            .build();
+
+    assertThat(event.hasAttribute(EventAttributes.CATALOG_NAME)).isTrue();
+    assertThat(event.hasAttribute(EventAttributes.TABLE_NAME)).isFalse();
+  }
+
+  @Test
+  void testAttributesReturnsUnmodifiableMap() {
+    PolarisEvent event =
+        PolarisEvent.builder(PolarisEventType.BEFORE_CREATE_TABLE, TEST_METADATA)
+            .attribute(EventAttributes.CATALOG_NAME, "my-catalog")
+            .build();
+
+    assertThat(event.attributes()).hasSize(1);
+    assertThat(event.attributes().get(EventAttributes.CATALOG_NAME)).isEqualTo("my-catalog");
+    assertThatThrownBy(() -> event.attributes().put(EventAttributes.TABLE_NAME, "table"))
+        .isInstanceOf(UnsupportedOperationException.class);
+  }
+
+  @Test
+  void testNullAttributeValueIsIgnored() {
+    PolarisEvent event =
+        PolarisEvent.builder(PolarisEventType.BEFORE_CREATE_TABLE, TEST_METADATA)
+            .attribute(EventAttributes.CATALOG_NAME, null)
+            .build();
+
+    assertThat(event.hasAttribute(EventAttributes.CATALOG_NAME)).isFalse();
+    assertThat(event.attributes()).isEmpty();
+  }
+
+  @Test
+  void testRequiredAttributeReturnsValue() {
+    PolarisEvent event =
+        PolarisEvent.builder(PolarisEventType.BEFORE_CREATE_TABLE, TEST_METADATA)
+            .attribute(EventAttributes.CATALOG_NAME, "my-catalog")
+            .build();
+
+    assertThat(event.requiredAttribute(EventAttributes.CATALOG_NAME)).isEqualTo("my-catalog");
+  }
+
+  @Test
+  void testRequiredAttributeThrowsForMissingKey() {
+    PolarisEvent event =
+        PolarisEvent.builder(PolarisEventType.BEFORE_CREATE_TABLE, TEST_METADATA).build();
+
+    assertThatThrownBy(() -> event.requiredAttribute(EventAttributes.CATALOG_NAME))
+        .isInstanceOf(IllegalStateException.class)
+        .hasMessageContaining("catalog_name")
```

Comment:
nit: Can we test for a more descriptive message? Not looking for exact string matching but something similar to "Required attribute" and/or "not found in event".